group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -0,0 +1,1040 @@
+         * watermark of the partition.
+         */
+        volatile long lastCommittedOffset;
+
+        /**
+         * Constructor.
+         *
+         * @param tp The topic partition of the coordinator.
+         */
+        private CoordinatorContext(
+            TopicPartition tp
+        ) {
+            this.tp = tp;
+            this.snapshotRegistry = new SnapshotRegistry(logContext);
+            this.deferredEventQueue = new DeferredEventQueue(logContext);
+            this.state = CoordinatorState.INITIAL;
+            this.epoch = -1;
+            this.lastWrittenOffset = 0L;
+            this.lastCommittedOffset = 0L;
+        }
+
+        /**
+         * Updates the last written offset. This also create a new snapshot
+         * in the snapshot registry.
+         *
+         * @param offset The new last written offset.
+         */
+        private void updateLastWrittenOffset(
+            long offset
+        ) {
+            if (offset <= lastWrittenOffset) {
+                throw new IllegalStateException("New last written offset " + offset + " of " + tp +
+                    " must be larger than " + lastWrittenOffset + ".");
+            }
+
+            log.debug("Update last written offset of {} to {}.", tp, offset);
+            lastWrittenOffset = offset;
+            snapshotRegistry.getOrCreateSnapshot(offset);
+        }
+
+        /**
+         * Reverts the last written offset. This also reverts the snapshot
+         * registry to this offset. All the changes applied after the offset
+         * are lost.
+         *
+         * @param offset The offset to revert to.
+         */
+        private void revertLastWrittenOffset(
+            long offset
+        ) {
+            if (offset > lastWrittenOffset) {
+                throw new IllegalStateException("New offset " + offset + " of " + tp +
+                    " must be smaller than " + lastWrittenOffset + ".");
+            }
+
+            log.debug("Revert last written offset of {} to {}.", tp, offset);
+            lastWrittenOffset = offset;
+            snapshotRegistry.revertToSnapshot(offset);
+        }
+
+        /**
+         * Updates the last committed offset. This completes all the deferred
+         * events waiting on this offset. This also cleanups all the snapshots
+         * prior to this offset.
+         *
+         * @param offset The new last committed offset.
+         */
+        private void updateLastCommittedOffset(
+            long offset
+        ) {
+            if (offset <= lastCommittedOffset) {
+                throw new IllegalStateException("New committed offset " + offset + " of " + tp +
+                    " must be larger than " + lastCommittedOffset + ".");
+            }
+
+            log.debug("Update committed offset of {} to {}.", tp, offset);
+            lastCommittedOffset = offset;
+            deferredEventQueue.completeUpTo(offset);
+            snapshotRegistry.deleteSnapshotsUpTo(offset);
```

Comment:
Hmm... All the events are already processed in a thread pool. We could defer those in another one but I don't really see the need for it for two reasons: 1) that would require to introduce locking because those are not thread safe; and 2) they should be quite fast operations.